airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py

```diff
@@ -22,47 +22,113 @@
+    Non_pagination = 0
+    Offset = 1
+    Metadata = 2
+
+
+class TooManyRequests(Exception):
+    """Too many requests"""
+
+
+class Client(BaseClient):
+    BASE_URL = "https://api.sendgrid.com/v3/"
 
-class Client:
     def __init__(self, apikey: str):
-        self._client = SendGridAPIClient(api_key=apikey)
-        self.ENTITY_MAP = {
-            "campaigns": self.campaigns,
-            "lists": self.lists,
-            "contacts": self.contacts,
-            "stats_automations": self.stats_automations,
-        }
+        self._headers = {"Authorization": f"Bearer {apikey}"}
+        super().__init__()
 
     def health_check(self):
-        try:
-            self._client.client.scopes.get()
+        resp_data = requests.get(f"{self.BASE_URL}scopes", headers=self._headers)
+        if resp_data.status_code == status_codes.OK:
             return True, None
-        except (UnauthorizedError, ForbiddenError) as err:
-            return False, err.args[0]
 
-    def get_streams(self):
-        streams = []
-        for schema, method in self.ENTITY_MAP.items():
-            raw_schema = json.loads(pkgutil.get_data(self.__class__.__module__.split(".")[0], f"schemas/{schema}.json"))
-            streams.append(AirbyteStream(name=schema, json_schema=raw_schema, supported_sync_modes=[SyncMode.full_refresh]))
-        return streams
+        return False, resp_data.json()["errors"][0]["message"]
+
+    @backoff.on_exception(backoff.expo, TooManyRequests, max_tries=6)
+    def _request(self, url: str, **kwargs) -> Response:
+        response = requests.get(url, headers=self._headers, **kwargs)
+        if response.status_code == status_codes.TOO_MANY_REQUESTS:
+            msg = f"Rate limit error: {response.json()}"
+            logger.error(msg)
+            raise TooManyRequests(msg)
+        elif response.status_code != status_codes.OK:
+            raise Exception(f"Unable to get data, error during request from url: {url}. Error: {response.json()}")
+        return response
+
+    def _read_data(self, stream_url: str, pagination: Pagination, key: str = None) -> Generator[dict, None, None]:
+        if pagination == Pagination.Non_pagination:
+            response = self._request(f"{self.BASE_URL}{stream_url}")
+            yield from response.json()[key] if key else response.json()
+        elif pagination == Pagination.Offset:
+            offset = 0
+            while True:
+                response = self._request(f"{self.BASE_URL}{stream_url}?limit={PAGE_SIZE}&offset={offset}")
+                stream_data = response.json()[key] if key else response.json()
+                yield from stream_data
+
+                if len(stream_data) < PAGE_SIZE:
+                    return
+                offset += PAGE_SIZE
+        elif pagination == Pagination.Metadata:
+            next_page = True
+            while next_page:
+                response = self._request(f"{self.BASE_URL}{stream_url}", params={"page_size": PAGE_SIZE})
```

Comment:
We don't see to be using `next_page` at all here...I think this isn't paging correctly

Reply:
And the truth is, I don't know how I could have missed this, it turns out to be an endless loop. Already fixed this bug.